javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,385 @@
+      last.next = null
+
+      _size -= 1
+      oldLast.value
+    }
+  }
+
+  def addFirst(e: E): Unit = {
+    val oldHead = head
+
+    head = new Node(e, next = oldHead)
+
+    _size += 1
+
+    if (oldHead ne null) {
+      oldHead.prev = head
+
+      if ((last eq null) && (oldHead.next eq null))
+        last = oldHead
+    }
+  }
+
+  def addLast(e: E): Unit = {
+    if (head eq null) addFirst(e)
+    else {
+      val oldLast =
+        if (last ne null) last
+        else if (head ne null) head
+        else null
+
+      last = new Node(e, prev = oldLast)
+
+      _size += 1
+
+      if (oldLast ne null)
+        oldLast.next = last
+    }
+  }
+
+  override def contains(o: Any): Boolean =
+    iterator().exists(_ === o)
+
+  override def size(): Int =
+    _size
+
+  override def add(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  override def remove(o: Any): Boolean =
+    _removeOccurrence(listIterator, o)
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator
+    val changed = iter.hasNext()
+    while (iter.hasNext())
+      addLast(iter.next())
+
+    changed
+  }
+
+  override def clear(): Unit = {
+    head = null
+    last = null
+    _size = 0
+  }
+
+  private def getNodeAt(index: Int): Node[E] = {
+    if (index == 0) head
+    else if (index == (size - 1)) last
+    else {
+      var current: Node[E] = null
+      if (index <= (size/2)) {
+        current = head
+        for (_ <- 0 until index) current = current.next
+      } else {
+        current = last
+        for (_ <- index until (size - 1)) current = current.prev
+      }
+      current
+    }
+  }
+
+  override def get(index: Int): E = {
+    checkIndexInBounds(index)
+    getNodeAt(index).value
+  }
+
+  override def set(index: Int, element: E): E = {
+    checkIndexInBounds(index)
+    val node = getNodeAt(index)
+    val oldValue = node.value
+    node.value = element
+    oldValue
+  }
+
+  private def addNode(next: Node[E], e: E): Unit = {
+    if (next eq head) addFirst(e)
+    else if (next eq null) addLast(e)
+    else {
+      var current = next
+
+      val node = new Node(e, prev = current.prev, next = current)
+      current.prev.next = node
+      current.prev = node
+
+      _size += 1
+    }
+  }
+
+  override def add(index: Int, element: E): Unit = {
+    checkIndexOnBounds(index)
+    if (index == 0) addFirst(element)
+    else if (index == size) addLast(element)
+    else addNode(getNodeAt(index), element)
+  }
+
+  private def removeNode(node: Node[E]): E = {
+    if (node eq head) removeFirst()
+    else if (node eq last) removeLast()
+    else {
+      var current = node
+
+      current.prev.next = current.next
+      current.next.prev = current.prev
+
+      _size -= 1
+
+      current.value
+    }
+  }
+
+  override def remove(index: Int): E = {
+    checkIndexInBounds(index)
+    if (index == 0) removeFirst()
+    else if (index == (size - 1)) removeLast()
+    else removeNode(getNodeAt(index))
```

Comment:
These cases are not needed. You can simply always call `removeNode(getNodeAt(index))`.